File: frontend/lib/agenda-route-rules.ts
export interface AgendaRotaBaseConfig {
  label: string;
  address: string;
  zip_code: string;
  lat: number | null;
  lng: number | null;
}

export interface AgendaRotaThresholdsConfig {
  nearby_anchor_max_travel_min: number;
  distant_clinic_min_travel_from_base_min: number;
  low_frequency_max_bookings_30d: number;
  max_insertion_detour_min: number;
  safe_margin_min: number;
}

export interface AgendaRotaOfferPolicyConfig {
  default_first_offer_days_ahead: number[];
  distant_low_frequency_first_offer_days_ahead: number[];
  allow_d2_if_anchor_exists: boolean;
  emergency_first_offer_days_ahead: number[];
}

export interface AgendaRotaRoutePolicyConfig {
  end_of_route_window_start: string;
  prefer_near_base_at_end_of_route: boolean;
  bonus_near_base_score: number;
  penalty_far_base_score: number;
  reject_clear_inefficiency: boolean;
}

export interface AgendaRotaFallbackPolicyConfig {
  suggest_alternative_slots_when_blocked: boolean;
  max_alternative_suggestions: number;
  allow_extra_slot_start_or_end_route_for_emergency: boolean;
}

export interface AgendaRotaRenderingPolicyConfig {
  use_custom_window: boolean;
  window_start: string;
  window_end: string;
  slot_interval_min: number;
}

export interface AgendaRotaClinicOverrideConfig {
  clinic_name: string;
  force_days_ahead: number[];
  prefer_only_when_anchor_exists: boolean;
  notes: string;
}

export interface AgendaRotaRegrasConfig {
  version: string;
  base: AgendaRotaBaseConfig;
  thresholds: AgendaRotaThresholdsConfig;
  offer_policy: AgendaRotaOfferPolicyConfig;
  route_policy: AgendaRotaRoutePolicyConfig;
  fallback_policy: AgendaRotaFallbackPolicyConfig;
  rendering_policy: AgendaRotaRenderingPolicyConfig;
  clinic_overrides: AgendaRotaClinicOverrideConfig[];
}

export const DEFAULT_AGENDA_ROTA_REGRAS: AgendaRotaRegrasConfig = {
  version: "1.0.0",
  base: {
    label: "Casa (base operacional)",
    address: "Av da Universidade, 1949",
    zip_code: "60020-180",
    lat: null,
    lng: null,
  },
  thresholds: {
    nearby_anchor_max_travel_min: 20,
    distant_clinic_min_travel_from_base_min: 35,
    low_frequency_max_bookings_30d: 3,
    max_insertion_detour_min: 25,
    safe_margin_min: 5,
  },
  offer_policy: {
    default_first_offer_days_ahead: [2],
    distant_low_frequency_first_offer_days_ahead: [3, 4],
    allow_d2_if_anchor_exists: true,
    emergency_first_offer_days_ahead: [1, 2],
  },
  route_policy: {
    end_of_route_window_start: "16:00",
    prefer_near_base_at_end_of_route: true,
    bonus_near_base_score: 15,
    penalty_far_base_score: 10,
    reject_clear_inefficiency: true,
  },
  fallback_policy: {
    suggest_alternative_slots_when_blocked: true,
    max_alternative_suggestions: 3,
    allow_extra_slot_start_or_end_route_for_emergency: true,
  },
  rendering_policy: {
    use_custom_window: false,
    window_start: "08:00",
    window_end: "18:00",
    slot_interval_min: 30,
  },
  clinic_overrides: [],
};

const normalizarInt = (value: unknown, fallback: number, min: number, max: number): number => {
  const parsed = Number.parseInt(String(value ?? fallback), 10);
  if (!Number.isFinite(parsed)) return fallback;
  return Math.max(min, Math.min(max, parsed));
};

const normalizarFloat = (value: unknown): number | null => {
  if (value == null || value === "") return null;
  const parsed = Number.parseFloat(String(value).replace(",", "."));
  if (!Number.isFinite(parsed)) return null;
  return parsed;
};

const normalizarBool = (value: unknown, fallback: boolean): boolean => {
  if (typeof value === "boolean") return value;
  if (typeof value === "number") return value !== 0;
  if (typeof value === "string") {
    const raw = value.trim().toLowerCase();
    if (["1", "true", "yes", "sim", "on"].includes(raw)) return true;
    if (["0", "false", "no", "nao", "não", "off"].includes(raw)) return false;
  }
  return fallback;
};

const normalizarHoraHHMM = (value: unknown, fallback: string): string => {
  const raw = String(value ?? "").trim();
  const match = raw.match(/^(\d{2}):(\d{2})$/);
  if (!match) return fallback;
  const hh = Number.parseInt(match[1], 10);
  const mm = Number.parseInt(match[2], 10);
  if (!Number.isFinite(hh) || !Number.isFinite(mm)) return fallback;
  if (hh < 0 || hh > 23 || mm < 0 || mm > 59) return fallback;
  return `${String(hh).padStart(2, "0")}:${String(mm).padStart(2, "0")}`;
};

const horaParaMinutos = (value: string): number => {
  const [hh, mm] = value.split(":").map((item) => Number.parseInt(item, 10));
  return (Number.isFinite(hh) ? hh : 0) * 60 + (Number.isFinite(mm) ? mm : 0);
};

export const normalizarDiasAFrente = (value: unknown, fallback: number[]): number[] => {
  const listaRaw = Array.isArray(value)
    ? value
    : String(value ?? "")
        .split(",")
        .map((item) => item.trim())
        .filter(Boolean);

  const dias = new Set<number>();
  for (const item of listaRaw) {
    const parsed = Number.parseInt(String(item), 10);
    if (!Number.isFinite(parsed)) continue;
    if (parsed < 0 || parsed > 30) continue;
    dias.add(parsed);
  }

  const ordenado = Array.from(dias).sort((a, b) => a - b);
  return ordenado.length > 0 ? ordenado : [...fallback];
};

const normalizarClinicOverrides = (value: unknown): AgendaRotaClinicOverrideConfig[] => {
  if (!Array.isArray(value)) return [];
  const normalized: AgendaRotaClinicOverrideConfig[] = [];

  for (const item of value) {
    if (!item || typeof item !== "object") continue;
    const row = item as Record<string, unknown>;
    const clinicName = String(row.clinic_name ?? "").trim();
    if (!clinicName) continue;
    normalized.push({
      clinic_name: clinicName,
      force_days_ahead: normalizarDiasAFrente(
        row.force_days_ahead,
        DEFAULT_AGENDA_ROTA_REGRAS.offer_policy.distant_low_frequency_first_offer_days_ahead
      ),
      prefer_only_when_anchor_exists: normalizarBool(row.prefer_only_when_anchor_exists, true),
      notes: String(row.notes ?? "").trim(),
    });
  }

  return normalized;
};

export const normalizarAgendaRotaRegras = (payload: unknown): AgendaRotaRegrasConfig => {
  const source = payload && typeof payload === "object" ? (payload as Record<string, unknown>) : {};
  const defaultCfg = DEFAULT_AGENDA_ROTA_REGRAS;

  const baseRaw =
    source.base && typeof source.base === "object" ? (source.base as Record<string, unknown>) : {};
  const thresholdsRaw =
    source.thresholds && typeof source.thresholds === "object"
      ? (source.thresholds as Record<string, unknown>)
      : {};
  const offerRaw =
    source.offer_policy && typeof source.offer_policy === "object"
      ? (source.offer_policy as Record<string, unknown>)
      : {};
  const routeRaw =
    source.route_policy && typeof source.route_policy === "object"
      ? (source.route_policy as Record<string, unknown>)
      : {};
  const fallbackRaw =
    source.fallback_policy && typeof source.fallback_policy === "object"
      ? (source.fallback_policy as Record<string, unknown>)
      : {};
  const renderingRaw =
    source.rendering_policy && typeof source.rendering_policy === "object"
      ? (source.rendering_policy as Record<string, unknown>)
      : {};

  const lat = normalizarFloat(baseRaw.lat);
  const lng = normalizarFloat(baseRaw.lng);
  const renderingStart = normalizarHoraHHMM(
    renderingRaw.window_start,
    defaultCfg.rendering_policy.window_start
  );
  const renderingEnd = normalizarHoraHHMM(
    renderingRaw.window_end,
    defaultCfg.rendering_policy.window_end
  );
  const renderingWindowValida = horaParaMinutos(renderingStart) < horaParaMinutos(renderingEnd);

  return {
    version: String(source.version ?? defaultCfg.version).trim() || defaultCfg.version,
    base: {
      label: String(baseRaw.label ?? defaultCfg.base.label).trim() || defaultCfg.base.label,
      address: String(baseRaw.address ?? defaultCfg.base.address).trim() || defaultCfg.base.address,
      zip_code: String(baseRaw.zip_code ?? defaultCfg.base.zip_code).trim() || defaultCfg.base.zip_code,
      lat: lat != null && lat >= -90 && lat <= 90 ? lat : null,
      lng: lng != null && lng >= -180 && lng <= 180 ? lng : null,
    },
    thresholds: {
      nearby_anchor_max_travel_min: normalizarInt(
        thresholdsRaw.nearby_anchor_max_travel_min,
        defaultCfg.thresholds.nearby_anchor_max_travel_min,
        1,
        240
      ),
      distant_clinic_min_travel_from_base_min: normalizarInt(
        thresholdsRaw.distant_clinic_min_travel_from_base_min,
        defaultCfg.thresholds.distant_clinic_min_travel_from_base_min,
        1,
        360
      ),
      low_frequency_max_bookings_30d: normalizarInt(
        thresholdsRaw.low_frequency_max_bookings_30d,
        defaultCfg.thresholds.low_frequency_max_bookings_30d,
        0,
        60
      ),
      max_insertion_detour_min: normalizarInt(
        thresholdsRaw.max_insertion_detour_min,
        defaultCfg.thresholds.max_insertion_detour_min,
        0,
        360
      ),
      safe_margin_min: normalizarInt(
        thresholdsRaw.safe_margin_min,
        defaultCfg.thresholds.safe_margin_min,
        0,
        120
      ),
    },
    offer_policy: {
      default_first_offer_days_ahead: normalizarDiasAFrente(
        offerRaw.default_first_offer_days_ahead,
        defaultCfg.offer_policy.default_first_offer_days_ahead
      ),
      distant_low_frequency_first_offer_days_ahead: normalizarDiasAFrente(
        offerRaw.distant_low_frequency_first_offer_days_ahead,
        defaultCfg.offer_policy.distant_low_frequency_first_offer_days_ahead
      ),
      allow_d2_if_anchor_exists: normalizarBool(
        offerRaw.allow_d2_if_anchor_exists,
        defaultCfg.offer_policy.allow_d2_if_anchor_exists
      ),
      emergency_first_offer_days_ahead: normalizarDiasAFrente(
        offerRaw.emergency_first_offer_days_ahead,
        defaultCfg.offer_policy.emergency_first_offer_days_ahead
      ),
    },
    route_policy: {
      end_of_route_window_start: normalizarHoraHHMM(
        routeRaw.end_of_route_window_start,
        defaultCfg.route_policy.end_of_route_window_start
      ),
      prefer_near_base_at_end_of_route: normalizarBool(
        routeRaw.prefer_near_base_at_end_of_route,
        defaultCfg.route_policy.prefer_near_base_at_end_of_route
      ),
      bonus_near_base_score: Math.abs(
        normalizarInt(routeRaw.bonus_near_base_score, defaultCfg.route_policy.bonus_near_base_score, -999, 999)
      ),
      penalty_far_base_score: normalizarInt(
        routeRaw.penalty_far_base_score,
        defaultCfg.route_policy.penalty_far_base_score,
        0,
        999
      ),
      reject_clear_inefficiency: normalizarBool(
        routeRaw.reject_clear_inefficiency,
        defaultCfg.route_policy.reject_clear_inefficiency
      ),
    },
    fallback_policy: {
      suggest_alternative_slots_when_blocked: normalizarBool(
        fallbackRaw.suggest_alternative_slots_when_blocked,
        defaultCfg.fallback_policy.suggest_alternative_slots_when_blocked
      ),
      max_alternative_suggestions: normalizarInt(
        fallbackRaw.max_alternative_suggestions,
        defaultCfg.fallback_policy.max_alternative_suggestions,
        1,
        20
      ),
      allow_extra_slot_start_or_end_route_for_emergency: normalizarBool(
        fallbackRaw.allow_extra_slot_start_or_end_route_for_emergency,
        defaultCfg.fallback_policy.allow_extra_slot_start_or_end_route_for_emergency
      ),
    },
    rendering_policy: {
      use_custom_window: normalizarBool(
        renderingRaw.use_custom_window,
        defaultCfg.rendering_policy.use_custom_window
      ),
      window_start: renderingWindowValida ? renderingStart : defaultCfg.rendering_policy.window_start,
      window_end: renderingWindowValida ? renderingEnd : defaultCfg.rendering_policy.window_end,
      slot_interval_min: normalizarInt(
        renderingRaw.slot_interval_min,
        defaultCfg.rendering_policy.slot_interval_min,
        5,
        120
      ),
    },
    clinic_overrides: normalizarClinicOverrides(source.clinic_overrides),
  };
};

export const formatarDiasAFrenteInput = (value: number[]): string => {
  return Array.isArray(value) ? value.join(", ") : "";
};

File: backend/app/core/agenda_route_rules.py
"""Helpers para configuracao de regras de rota da agenda."""
from __future__ import annotations

from copy import deepcopy
from typing import Any

from app.core.agenda_config import carregar_json

DEFAULT_AGENDA_ROTA_REGRAS = {
    "version": "1.0.0",
    "base": {
        "label": "Casa (base operacional)",
        "address": "Av da Universidade, 1949",
        "zip_code": "60020-180",
        "lat": None,
        "lng": None,
    },
    "thresholds": {
        "nearby_anchor_max_travel_min": 20,
        "distant_clinic_min_travel_from_base_min": 35,
        "low_frequency_max_bookings_30d": 3,
        "max_insertion_detour_min": 25,
        "safe_margin_min": 5,
    },
    "offer_policy": {
        "default_first_offer_days_ahead": [2],
        "distant_low_frequency_first_offer_days_ahead": [3, 4],
        "allow_d2_if_anchor_exists": True,
        "emergency_first_offer_days_ahead": [1, 2],
    },
    "route_policy": {
        "end_of_route_window_start": "16:00",
        "prefer_near_base_at_end_of_route": True,
        "bonus_near_base_score": 15,
        "penalty_far_base_score": 10,
        "reject_clear_inefficiency": True,
    },
    "fallback_policy": {
        "suggest_alternative_slots_when_blocked": True,
        "max_alternative_suggestions": 3,
        "allow_extra_slot_start_or_end_route_for_emergency": True,
    },
    "rendering_policy": {
        "use_custom_window": False,
        "window_start": "08:00",
        "window_end": "18:00",
        "slot_interval_min": 30,
    },
    "clinic_overrides": [],
}


def _normalizar_hora_hhmm(value: Any, fallback: str) -> str:
    raw = str(value or "").strip()
    if len(raw) != 5 or raw[2] != ":":
        return fallback

    hh = raw[:2]
    mm = raw[3:]
    if not (hh.isdigit() and mm.isdigit()):
        return fallback

    hora = int(hh)
    minuto = int(mm)
    if hora < 0 or hora > 23 or minuto < 0 or minuto > 59:
        return fallback
    return f"{hora:02d}:{minuto:02d}"


def _hora_em_minutos(value: str) -> int:
    hora_str, minuto_str = value.split(":")
    return int(hora_str) * 60 + int(minuto_str)


def _normalizar_int(value: Any, fallback: int, min_value: int, max_value: int) -> int:
    try:
        parsed = int(value)
    except (TypeError, ValueError):
        parsed = fallback
    return max(min_value, min(max_value, parsed))


def _normalizar_float(value: Any) -> float | None:
    if value in (None, "", "null"):
        return None
    try:
        parsed = float(value)
    except (TypeError, ValueError):
        return None
    return parsed


def _normalizar_bool(value: Any, fallback: bool) -> bool:
    if isinstance(value, bool):
        return value
    if isinstance(value, (int, float)):
        return value != 0
    if isinstance(value, str):
        raw = value.strip().lower()
        if raw in {"1", "true", "yes", "sim", "on"}:
            return True
        if raw in {"0", "false", "no", "nao", "não", "off"}:
            return False
    return fallback


def _normalizar_dias_a_frente(value: Any, fallback: list[int]) -> list[int]:
    raw_list = value if isinstance(value, list) else fallback
    dias: list[int] = []
    vistos: set[int] = set()
    for item in raw_list:
        try:
            dia = int(item)
        except (TypeError, ValueError):
            continue
        if dia < 0 or dia > 30:
            continue
        if dia in vistos:
            continue
        vistos.add(dia)
        dias.append(dia)
    dias.sort()
    return dias or list(fallback)


def normalizar_agenda_rota_regras(payload: Any) -> dict[str, Any]:
    source = payload if isinstance(payload, dict) else {}
    default = deepcopy(DEFAULT_AGENDA_ROTA_REGRAS)

    base_src = source.get("base") if isinstance(source.get("base"), dict) else {}
    base = {
        "label": str(base_src.get("label") or default["base"]["label"]).strip() or default["base"]["label"],
        "address": str(base_src.get("address") or default["base"]["address"]).strip() or default["base"]["address"],
        "zip_code": str(base_src.get("zip_code") or default["base"]["zip_code"]).strip() or default["base"]["zip_code"],
        "lat": _normalizar_float(base_src.get("lat")),
        "lng": _normalizar_float(base_src.get("lng")),
    }
    if base["lat"] is not None and not (-90.0 <= base["lat"] <= 90.0):
        base["lat"] = None
    if base["lng"] is not None and not (-180.0 <= base["lng"] <= 180.0):
        base["lng"] = None

    thresholds_src = source.get("thresholds") if isinstance(source.get("thresholds"), dict) else {}
    thresholds = {
        "nearby_anchor_max_travel_min": _normalizar_int(
            thresholds_src.get("nearby_anchor_max_travel_min"),
            default["thresholds"]["nearby_anchor_max_travel_min"],
            1,
            240,
        ),
        "distant_clinic_min_travel_from_base_min": _normalizar_int(
            thresholds_src.get("distant_clinic_min_travel_from_base_min"),
            default["thresholds"]["distant_clinic_min_travel_from_base_min"],
            1,
            360,
        ),
        "low_frequency_max_bookings_30d": _normalizar_int(
            thresholds_src.get("low_frequency_max_bookings_30d"),
            default["thresholds"]["low_frequency_max_bookings_30d"],
            0,
            60,
        ),
        "max_insertion_detour_min": _normalizar_int(
            thresholds_src.get("max_insertion_detour_min"),
            default["thresholds"]["max_insertion_detour_min"],
            0,
            360,
        ),
        "safe_margin_min": _normalizar_int(
            thresholds_src.get("safe_margin_min"),
            default["thresholds"]["safe_margin_min"],
            0,
            120,
        ),
    }

    offer_src = source.get("offer_policy") if isinstance(source.get("offer_policy"), dict) else {}
    offer_policy = {
        "default_first_offer_days_ahead": _normalizar_dias_a_frente(
            offer_src.get("default_first_offer_days_ahead"),
            default["offer_policy"]["default_first_offer_days_ahead"],
        ),
        "distant_low_frequency_first_offer_days_ahead": _normalizar_dias_a_frente(
            offer_src.get("distant_low_frequency_first_offer_days_ahead"),
            default["offer_policy"]["distant_low_frequency_first_offer_days_ahead"],
        ),
        "allow_d2_if_anchor_exists": _normalizar_bool(
            offer_src.get("allow_d2_if_anchor_exists"),
            default["offer_policy"]["allow_d2_if_anchor_exists"],
        ),
        "emergency_first_offer_days_ahead": _normalizar_dias_a_frente(
            offer_src.get("emergency_first_offer_days_ahead"),
            default["offer_policy"]["emergency_first_offer_days_ahead"],
        ),
    }

    route_src = source.get("route_policy") if isinstance(source.get("route_policy"), dict) else {}
    raw_bonus = _normalizar_int(
        route_src.get("bonus_near_base_score"),
        default["route_policy"]["bonus_near_base_score"],
        -999,
        999,
    )
    route_policy = {
        "end_of_route_window_start": _normalizar_hora_hhmm(
            route_src.get("end_of_route_window_start"),
            default["route_policy"]["end_of_route_window_start"],
        ),
        "prefer_near_base_at_end_of_route": _normalizar_bool(
            route_src.get("prefer_near_base_at_end_of_route"),
            default["route_policy"]["prefer_near_base_at_end_of_route"],
        ),
        "bonus_near_base_score": abs(raw_bonus),
        "penalty_far_base_score": _normalizar_int(
            route_src.get("penalty_far_base_score"),
            default["route_policy"]["penalty_far_base_score"],
            0,
            999,
        ),
        "reject_clear_inefficiency": _normalizar_bool(
            route_src.get("reject_clear_inefficiency"),
            default["route_policy"]["reject_clear_inefficiency"],
        ),
    }

    fallback_src = source.get("fallback_policy") if isinstance(source.get("fallback_policy"), dict) else {}
    fallback_policy = {
        "suggest_alternative_slots_when_blocked": _normalizar_bool(
            fallback_src.get("suggest_alternative_slots_when_blocked"),
            default["fallback_policy"]["suggest_alternative_slots_when_blocked"],
        ),
        "max_alternative_suggestions": _normalizar_int(
            fallback_src.get("max_alternative_suggestions"),
            default["fallback_policy"]["max_alternative_suggestions"],
            1,
            20,
        ),
        "allow_extra_slot_start_or_end_route_for_emergency": _normalizar_bool(
            fallback_src.get("allow_extra_slot_start_or_end_route_for_emergency"),
            default["fallback_policy"]["allow_extra_slot_start_or_end_route_for_emergency"],
        ),
    }

    rendering_src = source.get("rendering_policy") if isinstance(source.get("rendering_policy"), dict) else {}
    rendering_policy = {
        "use_custom_window": _normalizar_bool(
            rendering_src.get("use_custom_window"),
            default["rendering_policy"]["use_custom_window"],
        ),
        "window_start": _normalizar_hora_hhmm(
            rendering_src.get("window_start"),
            default["rendering_policy"]["window_start"],
        ),
        "window_end": _normalizar_hora_hhmm(
            rendering_src.get("window_end"),
            default["rendering_policy"]["window_end"],
        ),
        "slot_interval_min": _normalizar_int(
            rendering_src.get("slot_interval_min"),
            default["rendering_policy"]["slot_interval_min"],
            5,
            120,
        ),
    }
    if _hora_em_minutos(rendering_policy["window_start"]) >= _hora_em_minutos(rendering_policy["window_end"]):
        rendering_policy["window_start"] = default["rendering_policy"]["window_start"]
        rendering_policy["window_end"] = default["rendering_policy"]["window_end"]

    overrides_src = source.get("clinic_overrides") if isinstance(source.get("clinic_overrides"), list) else []
    clinic_overrides: list[dict[str, Any]] = []
    for item in overrides_src:
        if not isinstance(item, dict):
            continue
        clinic_name = str(item.get("clinic_name") or "").strip()
        if not clinic_name:
            continue
        clinic_overrides.append(
            {
                "clinic_name": clinic_name,
                "force_days_ahead": _normalizar_dias_a_frente(
                    item.get("force_days_ahead"),
                    offer_policy["distant_low_frequency_first_offer_days_ahead"],
                ),
                "prefer_only_when_anchor_exists": _normalizar_bool(
                    item.get("prefer_only_when_anchor_exists"),
                    True,
                ),
                "notes": str(item.get("notes") or "").strip(),
            }
        )

    return {
        "version": str(source.get("version") or default["version"]).strip() or default["version"],
        "base": base,
        "thresholds": thresholds,
        "offer_policy": offer_policy,
        "route_policy": route_policy,
        "fallback_policy": fallback_policy,
        "rendering_policy": rendering_policy,
        "clinic_overrides": clinic_overrides,
    }


def carregar_agenda_rota_regras(raw: Any) -> dict[str, Any]:
    payload = carregar_json(raw)
    return normalizar_agenda_rota_regras(payload)
